oauth_samples/go/oauth.go

```
@@ -0,0 +1,155 @@
+		fmt.Println("Erorr reading the code you entered")
+		return readerError
+	}
+
+	requestToken(strings.TrimSuffix(text, "\n"))
+
+	return nil
+}
+
+func requestToken(code string) error {
+	fmt.Println("Requesting access token")
+
+	// Step 2, 3 - turn the authorization code into a access token, etc
+	token, err := exchangeForToken(code)
+
+	if err != nil {
+		fmt.Println("Invalid token response")
+		return err
+	}
+
+	// We can now use the access_token as much as we want to access protected resources.
+	return makeAPICall(token)
+}
+
+func makeAPICall(token string) error {
+	var hc = http.DefaultClient
+	var bearer = "Bearer " + token
+
+	request, requestError := http.NewRequest("GET", apiURL, nil)
+
+	if requestError != nil {
+		fmt.Println("Error creating request")
+		return requestError
+	}
+
+	request.Header.Add("Authorization", bearer)
+	apiResp, responseError := hc.Do(request)
+
+	if responseError != nil {
+		fmt.Println("Error making Authorization request")
+		return responseError
+	}
+
+	if apiResp.StatusCode != http.StatusOK {
+		return fmt.Errorf("Unexpected respose %q", apiResp.Status)
+	}
+
+	if ct := apiResp.Header.Get("Content-Type"); ct != "application/json" {
+		return fmt.Errorf("Unsupported content-type: %q", ct)
+	}
+
+	defer apiResp.Body.Close()
+
+	_, readAllError := io.Copy(os.Stdout, apiResp.Body)
+	return readAllError
+}
+
+func exchangeForToken(code string) (string, error) {
+	authData := url.Values{
+		"grant_type":    {"authorization_code"},
+		"code":          {code},
+		"redirect_uri":  {callbackURL},
+		"client_id":     {clientID},
+		"client_secret": {clientSecret},
+	}
+
+	authResp, err := http.PostForm(tokenURL, authData)
+
+	if authResp == nil {
+		fmt.Println("error: ", err)
+		return "", err
+	}
+
+	fmt.Println("Access token requested")
+
+	var result oauth2.Token
+
+	body, readAllError := ioutil.ReadAll(authResp.Body)
+
+	if readAllError != nil {
+		fmt.Println("Error reading the API call response")
+		return "", readAllError
```

Better to return an annotated error here instead of print + return.